lib/elixir/src/elixir_clauses.erl

```diff
@@ -40,8 +40,18 @@ guard({'when', Meta, [Left, Right]}, E) ->
   {ELeft, EL}  = guard(Left, E),
   {ERight, ER} = guard(Right, EL),
   {{'when', Meta, [ELeft, ERight]}, ER};
-guard(Other, E) ->
-  elixir_expand:expand(Other, E).
+guard(Guard, E) ->
+  warn_zero_length_guard(Guard, E),
+  elixir_expand:expand(Guard, E).
+
+warn_zero_length_guard({'==', Meta, [{length, _, [{Var, _, _}]}, 0]}, E) ->
```

Comment:
Do not assume the argument for the function is a var. Have it written as `{length, _, [Arg]}` and then use `'Elixir.Macro':to_string(Arg)`to convert said Arg to its string representation.